readme.md
BecklynSecurityBundle
=====================

Security related helpers for symfony apps.


Installation
------------

```bash
composer require becklyn/security-bundle
```

and register the bundle in the `AppKernel`.


Configuration
-------------


Currently there is nothing else to do than activating this bundle.


CHANGELOG.md
2.1.0
=====

*   (feature) Add support for PHP 8.
*   (internal) Fix Symfony deprecations.
*   (internal) Replace TravisCI with GitHub Actions.
*   (improvement) Add missing property types and return types.
*   (internal) Remove support for Symfony 4.4.
*   (improvement) Add support for Symfony 6.


2.0.0
=====

*   Require Symfony 5.


1.0.0
=====

Initial Release `\o/`


Tests/Html/HtmlNonceInjectorTest.php
<?php

namespace Tests\Becklyn\SecurityBundle\Tests\Html;

use Becklyn\SecurityBundle\Html\HtmlNonceInjector;
use PHPUnit\Framework\TestCase;
use Psr\Log\NullLogger;


class HtmlNonceInjectorTest extends TestCase
{
    public function testInjection () : void
    {
        $html = '<html><body><p>Test</p></body></html>';
        $injector = new HtmlNonceInjector(new NullLogger());

        $withInjection = $injector->injectNonce($html);
        self::assertStringMatchesFormat('<html><body><p>Test</p>%s</body></html>', $withInjection);
    }
}


tests/Mitigation/BreachResponseListenerTest.php
<?php

namespace Tests\Becklyn\SecurityBundle\Tests\Mitigation\Breach;

use Becklyn\SecurityBundle\Html\HtmlNonceInjector;
use Becklyn\SecurityBundle\Mitigation\Breach\BreachResponseListener;
use PHPUnit\Framework\MockObject\MockObject;
use PHPUnit\Framework\TestCase;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpKernel\Event\ResponseEvent;
use Symfony\Component\HttpKernel\HttpKernelInterface;
use Symfony\Component\HttpKernel\KernelInterface;

class BreachResponseListenerTest extends TestCase
{
    /**
     * Builds all services
     *
     * @param HtmlNonceInjector $nonceInjector
     * @param bool              $isMasterRequest
     * @param bool              $isSecure
     * @return array
     */
    private function buildServices (HtmlNonceInjector $nonceInjector, bool $isMasterRequest, bool $isSecure) : array
    {
        $response = $this->createMock(Response::class);
        $request = $this->createMock(Request::class);

        $event = new ResponseEvent(
            $this->createMock(KernelInterface::class),
            $request,
            $isMasterRequest ? HttpKernelInterface::MASTER_REQUEST : HttpKernelInterface::SUB_REQUEST,
            $response
        );

        if ($isMasterRequest)
        {
            $request
                ->expects(self::once())
                ->method("isSecure")
                ->willReturn($isSecure);
        }

        return [
            new BreachResponseListener($nonceInjector),
            $event,
            $response
        ];
    }


    /**
     * @return array
     */
    public function dataProviderInvalidMasterSecurePermutations () : array
    {
        return [
            [true, false],
            [false, true],
            [false, false],
        ];
    }


    /**
     * Tests all invalid cases, where the listener shouldn't run
     *
     * @dataProvider dataProviderInvalidMasterSecurePermutations
     *
     * @param bool $isMasterRequest
     * @param bool $isSecure
     */
    public function testInvalidMasterSecurePermutations (bool $isMasterRequest, bool $isSecure) : void
    {
        $nonceInjector = $this->getMockBuilder(HtmlNonceInjector::class)
            ->disableOriginalConstructor()
            ->getMock();

        /**
         * @type BreachResponseListener $listener
         * @type \PHPUnit_Framework_MockObject_MockObject $event
         */
        [$listener, $event] = $this->buildServices($nonceInjector, $isMasterRequest, $isSecure);

        $nonceInjector
            ->expects(self::never())
            ->method("injectNonce");

        $listener->onResponse($event);
    }


    /**
     * Tests the correct interaction of the listener with the response + nonce injection
     */
    public function testEventRegistration ()
    {
        $nonceInjector = $this->getMockBuilder(HtmlNonceInjector::class)
            ->disableOriginalConstructor()
            ->getMock();

        /**
         * @type BreachResponseListener $listener
         * @type ResponseEvent $event
         * @type MockObject $response
         */
        [$listener, $event, $response] = $this->buildServices($nonceInjector, true, true);

        $rawHtml = '<html>content</html>';
        $injectedHtml = '<html>content<!-- injected --></html>';

        $nonceInjector
            ->expects(self::once())
            ->method("injectNonce")
            ->with($rawHtml)
            ->willReturn($injectedHtml);

        $response
            ->expects(self::once())
            ->method("getContent")
            ->willReturn($rawHtml);

        $response
            ->expects(self::once())
            ->method("setContent")
            ->with($injectedHtml);

        $listener->onResponse($event);
    }
}


Tests/Mitigation/Breach/BreachResponseListenerTest.php
<?php

namespace Tests\Becklyn\SecurityBundle\Tests\Mitigation\Breach;

use Becklyn\SecurityBundle\Html\HtmlNonceInjector;
use Becklyn\SecurityBundle\Mitigation\Breach\BreachResponseListener;
use PHPUnit\Framework\TestCase;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpKernel\Event\FilterResponseEvent;


class BreachResponseListenerTest extends TestCase
{
    /**
     * Builds all services
     *
     * @param HtmlNonceInjector $nonceInjector
     * @param bool              $isMasterRequest
     * @param bool              $isSecure
     * @return array
     */
    private function buildServices (HtmlNonceInjector $nonceInjector, bool $isMasterRequest, bool $isSecure) : array
    {
        $event = $this->getMockBuilder(FilterResponseEvent::class)
            ->disableOriginalConstructor()
            ->getMock();

        $request = $this->getMockBuilder(Request::class)
            ->disableOriginalConstructor()
            ->getMock();

        if ($isMasterRequest)
        {
            $request
                ->expects(self::once())
                ->method("isSecure")
                ->willReturn($isSecure);
        }

        $response = $this->getMockBuilder(Response::class)
            ->disableOriginalConstructor()
            ->getMock();

        $event
            ->expects(self::once())
            ->method("isMasterRequest")
            ->willReturn($isMasterRequest);

        $event
            ->method("getRequest")
            ->willReturn($request);

        $event
            ->method("getResponse")
            ->willReturn($response);

        return [
            new BreachResponseListener($nonceInjector),
            $event,
            $response
        ];
    }


    /**
     * @return array
     */
    public function dataProviderInvalidMasterSecurePermutations () : array
    {
        return [
            [true, false],
            [false, true],
            [false, false],
        ];
    }


    /**
     * Tests all invalid cases, where the listener shouldn't run
     *
     * @dataProvider dataProviderInvalidMasterSecurePermutations
     *
     * @param bool $isMasterRequest
     * @param bool $isSecure
     */
    public function testInvalidMasterSecurePermutations (bool $isMasterRequest, bool $isSecure) : void
    {
        $nonceInjector = $this->getMockBuilder(HtmlNonceInjector::class)
            ->disableOriginalConstructor()
            ->getMock();

        /**
         * @type BreachResponseListener $listener
         * @type \PHPUnit_Framework_MockObject_MockObject $event
         */
        [$listener, $event] = $this->buildServices($nonceInjector, $isMasterRequest, $isSecure);

        $nonceInjector
            ->expects(self::never())
            ->method("injectNonce");

        $listener->onResponse($event);
    }


    /**
     * Tests the correct interaction of the listener with the response + nonce injection
     */
    public function testEventRegistration ()
    {
        $nonceInjector = $this->getMockBuilder(HtmlNonceInjector::class)
            ->disableOriginalConstructor()
            ->getMock();

        /**
         * @type BreachResponseListener $listener
         * @type \PHPUnit_Framework_MockObject_MockObject $event
         * @type \PHPUnit_Framework_MockObject_MockObject $response
         */
        [$listener, $event, $response] = $this->buildServices($nonceInjector, true, true);

        $rawHtml = '<html>content</html>';
        $injectedHtml = '<html>content<!-- injected --></html>';


        $nonceInjector
            ->expects(self::once())
            ->method("injectNonce")
            ->with($rawHtml)
            ->willReturn($injectedHtml);

        $response
            ->expects(self::once())
            ->method("getContent")
            ->willReturn($rawHtml);

        $response
            ->expects(self::once())
            ->method("setContent")
            ->with($injectedHtml);

        $listener->onResponse($event);
    }
}


src/Html/HtmlNonceInjector.php
<?php declare(strict_types=1);

namespace Becklyn\SecurityBundle\Html;

use Psr\Log\LoggerInterface;

class HtmlNonceInjector
{
    private LoggerInterface $logger;


    public function __construct (LoggerInterface $logger)
    {
        $this->logger = $logger;
    }


    /**
     * Injects a nonce in the HTML, if it is a full HTML response (i.e. if there is a closing body tag).
     */
    public function injectNonce (string $html) : string
    {
        try
        {
            $pos = \strripos($html, '</body>');

            if (false === $pos)
            {
                return $html;
            }

            $randomData = \random_bytes(25);
            $nonce = \substr(\base64_encode($randomData), 0, \random_int(1, 32));
            $nonceSnippet = "<!-- {$nonce} -->";

            return \substr($html, 0, $pos) . $nonceSnippet . \substr($html, $pos);
        }
        catch (\Exception $e)
        {
            $this->logger->error("Could not inject nonce as random data generation failed.", [
                "innerException" => $e,
            ]);

            return $html;
        }
    }
}


src/Mitigation/Breach/BreachResponseListener.php
<?php declare(strict_types=1);

namespace Becklyn\SecurityBundle\Mitigation\Breach;

use Becklyn\SecurityBundle\Html\HtmlNonceInjector;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\HttpFoundation\BinaryFileResponse;
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\HttpFoundation\RedirectResponse;
use Symfony\Component\HttpFoundation\StreamedResponse;
use Symfony\Component\HttpKernel\Event\ResponseEvent;
use Symfony\Component\HttpKernel\KernelEvents;

/**
 * Tries to mitigate the BREACH SSL + deflate attack.
 *
 * HTTPS is vulnerable against a length guessing attack,
 * if there is user-directed content in a web page with HTTPS + gzip.
 *
 * As it is no option to deactivate either one of them, the response is padded with random content.
 */
class BreachResponseListener implements EventSubscriberInterface
{
    private HtmlNonceInjector $nonceInjector;


    public function __construct (HtmlNonceInjector $nonceInjector)
    {
        $this->nonceInjector = $nonceInjector;
    }


    public function onResponse (ResponseEvent $event) : void
    {
        if (!$event->isMainRequest() || !$event->getRequest()->isSecure())
        {
            return;
        }

        $response = $event->getResponse();

        switch (true)
        {
            // don't modify unsupported responses
            case $response instanceof BinaryFileResponse:
            case $response instanceof JsonResponse:
            case $response instanceof StreamedResponse:
            case $response instanceof RedirectResponse:
                return;

            default:
                $response->setContent(
                    $this->nonceInjector->injectNonce($response->getContent())
                );
                break;
        }
    }


    /**
     * @inheritdoc
     */
    public static function getSubscribedEvents () : array
    {
        return [
            KernelEvents::RESPONSE => "onResponse",
        ];
    }
}


EventListener/BreachPreventionResponseSubscriber.php
<?php

namespace Becklyn\SecurityBundle\EventListener;

use Psr\Log\LoggerInterface;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
use Symfony\Component\HttpKernel\KernelEvents;


/**
 * Tries to mitigate the BREACH SSL + deflate attack.
 *
 * HTTPS is vulnerable against a length guessing attack,
 * if there is user-directed content in a web page with HTTPS + gzip.
 *
 * As it is no option to deactivate either one of them, the response is padded with random content.
 */
class BreachPreventionResponseSubscriber implements EventSubscriberInterface
{
    /**
     * @var LoggerInterface
     */
    private $logger;


    /**
     * @param LoggerInterface $logger
     */
    public function __construct (LoggerInterface $logger)
    {
        $this->logger = $logger;
    }


    /**
     * @param FilterResponseEvent $event
     */
    public function onResponse (FilterResponseEvent $event)
    {
        if (!$event->isMasterRequest() || !$event->getRequest()->isSecure())
        {
            return;
        }

        $this->injectNonce($event->getResponse());
    }


    /**
     * Injects the nonce in the response
     *
     * @param Response $response
     */
    private function injectNonce (Response $response)
    {
        try
        {
            $content = $response->getContent();
            $pos = \strripos($content, '</body>');

            if (false === $pos)
            {
                return;
            }

            $randomData = random_bytes(25);
            $nonce = substr(base64_encode($randomData), 0, random_int(1, 32));
            $nonceSnippet = "<!-- {$nonce} -->";

            $content = substr($content, 0, $pos) . $nonceSnippet . substr($content, $pos);
            $response->setContent($content);
        }
        catch (\Exception $e)
        {
            $this->logger->error("Could not inject nonce as random data generation failed.", [
                "innerException" => $e,
            ]);
        }
    }


    /**
     * @inheritdoc
     */
    public static function getSubscribedEvents ()
    {
        return [
            KernelEvents::RESPONSE => "onResponse",
        ];
    }
}
